### PR TITLE
Update attribute setter/getters in various ways

### LINKED ISSUE #186: Should promise-returning attributes reject in case of wrong this value?
When a promise-returning attribute getter is called on a wrong this value, it is expected to throw a type error. Promise-returning operations on the other side will return a rejected promise.

I am wondering why attribute getters are not also rejecting.
Although not WebIDL based, https://streams.spec.whatwg.org/#default-reader-closed is doing so for instance.

### PR DESCRIPTION
**Pull request by domenic**
- Fixes getters on the global when no explicit `this` is passed, similarly to what was already done for operations in a2b4599631fe0fa254a7a4993dd7f00b74c8ad20. Closes https://www.w3.org/Bugs/Public/show_bug.cgi?id=29421.
- Make promise getters reject in case of an error. Disallow promise types on attributes that will generate setters. Fixes #186 and fixes https://www.w3.org/Bugs/Public/show_bug.cgi?id=25048.
- Fixes https://www.w3.org/Bugs/Public/show_bug.cgi?id=27905 by more clearly specifying the target (as was done previously for operations).
- Updated to use modern ES Get/Set abstract operations instead of [[Get]] and [[Put]].
- Also made minor fixes to the operations definition, e.g. wrapping it in `<div algorithm>` and remembering to actually return F.

***
[Preview](https://cdn.rawgit.com/heycam/webidl/9962a71/index.html) | [Diff w/ current ED](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fheycam.github.io%2Fwebidl%2F&doc2=https%3A%2F%2Fcdn.rawgit.com%2Fheycam%2Fwebidl%2F9962a71%2Findex.html) | [Diff w/ base](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fcdn.rawgit.com%2Fheycam%2Fwebidl%2F0c5f48b%2Findex.html&doc2=https%3A%2F%2Fcdn.rawgit.com%2Fheycam%2Fwebidl%2F9962a71%2Findex.html)

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -10724,130 +10728,141 @@ The characteristics of this property are as follows:
...
+            1.  If |attribute| is not a [=static attribute=]:
+                1.  If the <emu-val>this</emu-val> value is <emu-val>null</emu-val> or
+                    <emu-val>undefined</emu-val>, set |O| to |realm|'s [=Realm/global object=].
+                    (This will subsequently cause a <emu-val>TypeError</emu-val> in a few steps, if
+                    the global object does not implement |target| and [{{LenientThis}}] is not
+                    specified.)
+                    <!-- https://www.w3.org/Bugs/Public/show_bug.cgi?id=18547#c9 -->
+                1.  Otherwise, set |O| to the <emu-val>this</emu-val> value.
+                1.  If |O| is a [=platform object=], then [=perform a security check=], passing |O|,
+                    |realm|, |id|, and "method".
```
Comment (bzbarsky) [inline index.bs]: "getter", not "method", right?

File: index.bs [hunk 2]
```diff
@@ -10724,130 +10728,141 @@ The characteristics of this property are as follows:
...
+                of the inherited attribute, if this attribute is declared to
+                [=inherit its getter=]), on |O| if |O| is not <emu-val>null</emu-val>.
+            1.  Return the result of [=converted to an ECMAScript value|converting=] |R| to an
+                ECMAScript value of the type |attribute| is declared as.
+
+        And then, if an exception was thrown:
+
+        1.  If |attribute|'s type is a [=promise type=], then:
+            1.  Let |reject| be the initial value of [=%Promise%=].reject.
+            1.  Return the result of calling |reject| with [=%Promise%=] as the
```
Comment (bzbarsky) [inline index.bs]: Did we ever decide how this should interact with [SameObject]?  Seems to me like we should just disallow [SameObject] on Promise-returning attributes, or change the definition of [SameObject] or something; the current definition of [SameObject] just would not match this new behavior.

I'm still not 100% sure we want to make this behavior change, for what it's worth.  What do UAs do in practice?

File: index.bs [hunk 3]
```diff
@@ -10724,130 +10728,141 @@ The characteristics of this property are as follows:
...
+        1.  If |attribute| is not a [=static attribute=]:
+            1.  If the <emu-val>this</emu-val> value is <emu-val>null</emu-val> or
+                <emu-val>undefined</emu-val>, set |O| to |realm|'s [=Realm/global object=].
+                (This will subsequently cause a <emu-val>TypeError</emu-val> in a few steps, if
+                the global object does not implement |target| and [{{LenientThis}}] is not
+                specified.)
+                <!-- https://www.w3.org/Bugs/Public/show_bug.cgi?id=18547#c9 -->
+            1.  Otherwise, set |O| to the <emu-val>this</emu-val> value.
+            1.  If |O| is a [=platform object=], then [=perform a security check=], passing |O|,
+                |realm|, |id|, and "method".
```
Comment (bzbarsky) [inline index.bs]: "setter", not "method".

File: index.bs [hunk 4]
```diff
@@ -10902,53 +10917,56 @@ For [=namespaces=], the properties corresponding to each declared operation are
...
-    |steps|, the [=%FunctionPrototype%=] of |realm|).
-1.  Perform [=!=] [=SetFunctionName=](|F|, |id|).
-1.  Let |S| be the [=effective overload set=] for [=regular operations=] (if |op| is a regular
-    operation) or for [=static operations=] (if |op| is a static operation) with [=identifier=] |id|
-    on |target| and with argument count 0.
-1.  Let |length| be the length of the shortest argument list in the entries in |S|.
-1.  Perform [=!=] [=DefinePropertyOrThrow=](|F|, "length",
-    PropertyDescriptor<span class="prop-desc">{\[[Value]]: |length|, \[[Writable]]: <emu-val>false</emu-val>, \[[Enumerable]]: <emu-val>false</emu-val>, \[[Configurable]]: <emu-val>true</emu-val>}</span>).
+<div algorithm>
+    To <dfn id="dfn-create-operation-function" lt="creating an operation function">create an operation function</dfn>,
```
Comment (bzbarsky) [inline index.bs]: Wish this reindenting were a separate commit....

File: index.bs [hunk 5]
```diff
@@ -1385,6 +1385,10 @@ ignored or an exception is thrown.
     };
 </pre>
 
+The type of the attribute that is not [=read only=] must not be a [=promise type=]. Additionally,
```
Comment (domenic) [inline index.bs]: Fixed

File: index.bs (hunk 1 above)
Comment (domenic) [inline index.bs]: Fixed

File: index.bs (hunk 2 above)
Comment (domenic) [inline index.bs]: Fixed [SameObject] to reflect this.

In practice UAs are waiting for us to make a decision here; they throw the error per the current spec but are uncomfortable with doing so.

File: index.bs (hunk 4 above)
Comment (domenic) [inline index.bs]: Separated out into #232, but this PR for now is on top of that so it's included here.

File: index.bs (hunk 2 above)
Comment (bzbarsky) [inline index.bs]: > Fixed [SameObject] to reflect this.

I would really rather we just disallowed using [SameObject] on Promise-valued attributes if we take this route, given the weird behavior.  If we do allow it, UAs have to start doing something weird themselves, because they can no longer trust [SameObject] annotations...

In particular, it's not obvious that these changes to [SameObject] are compatible with #212

> they throw the error per the current spec but are uncomfortable with doing so.

I don't think Gecko is uncomfortable.  It sounds like Blink is?  Do you have references for the other UAs, by any chance?

Comment (domenic) [inline index.bs]: > I would really rather we just disallowed using [SameObject] on Promise-valued attributes if we take this route, given the weird behavior.

I really don't think it's that weird. It's about the exception case. You don't require the same exception objects be thrown each time, so similarly, you shouldn't require the same promise object be returned each time.

> Do you have references for the other UAs, by any chance?

WebKit: https://github.com/heycam/webidl/issues/186

Comment (bzbarsky) [inline index.bs]: > I really don't think it's that weird. It's about the exception case.

The problem is that unlike sync functions, which have a return value and a possible exception which are exposed to the caller in different ways, Promise-returning functions use a return value to represent both things.  This makes [SameObject] a pretty confusing thing to start with: is it referring to the Promise object returned, or to its resolution value in the non-exception case, for example?  If we really reason by analogy to sync functions, it should be referring to the resolution value...  but in practice people and the spec use it to mean the actual Promise return value.  Which doesn't even have an explicit equivalent in the sync function case in JS; it would be a continuation in languages that have such things.  So any reasoning by analogy here is broken right now anyway.

It's worth taking a step back for a second and asking what the point of [SameObject] really is and why we have it at all, then trying to figure out what it means to map that concept to promise-returning getters.  Assuming that's really possible.

> WebKit: #186

Thank you.  And Edge?

Comment (domenic) [inline index.bs]: > It's worth taking a step back for a second and asking what the point of [SameObject] really is and why we have it at all, then trying to figure out what it means to map that concept to promise-returning getters. Assuming that's really possible.

Right. I perhaps am not the best person for this, since I don't like these documentation-only extended attributes. But I tried to put myself in the shoes of someone who does, and an interpretation of "returns the same object in non-exceptional cases" seemed like a good thing to document.

> Thank you. And Edge?

I haven't seen any preferences from them.

Comment (bzbarsky) [inline index.bs]: OK.  Well, as another point of view, with my JIT implementor hat on, after this change all Promise-returning cases would be non-exceptional cases (because from a JIT's point of view they are), so my JIT would have to know that [SameObject] doesn't mean "the same object" in Promise-returning cases and hence can't be optimized the same way as other [SameObject] annotations...

File: index.bs [hunk 6]
```diff
@@ -1374,6 +1374,10 @@ ignored or an exception is thrown.
     };
 </pre>
 
+The type of an attribute that is not [=read only=] must not be a [=promise type=]. Additionally,
+the type of an attribute with any of the [=extended attributes=] [{{LenientThis}}],
+[{{PutForwards}}], [{{Replaceable}}], or [{{SameObject}}] must not be a [=promise type=].
```
Comment (tobie) [inline index.bs]: I thought there were requirements for [{{SameObject}}] promise type attributes. E.g. for https://drafts.csswg.org/css-font-loading/#font-face-set-ready, no?

Comment (domenic) [inline index.bs]: Yes, but @bzbarsky prefers we don't use them, as discussed upthread.

File: index.bs [hunk 7]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
-        1.  Let |V| be the result of [=converted to an ECMAScript value|converting=]
-            |idlValue| to an ECMAScript value.
-        1.  Return |V|.
-    </div>
+    *   |G| is the [=attribute getter=] created given the attribute, the interface (or the interface
+        it's being mixed in to, if the interface is actually a mixin), and the [=relevant Realm=] of
+        the object that is the location of the property; and
+    *   |S| is the [=attribute setter=] created given the attribute, the interface (or the interface
+        it's being mixed in to, if the interface is actually a mixin), and the [=relevant Realm=] of
+        the object that is the location of the property.
```
Comment (tobie) [inline index.bs]: I'm aware this is purposefully copied from the text in the operations binding, but: s/the object that is the location of the property/the object that contains the property/ would read better.

Maybe change it everywhere? Or not at all for now also works. :)

File: index.bs [hunk 8]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
+    1. Return |F|.
+</div>
 
 Note: Although there is only a single property for an IDL attribute, since
 accessor property getters and setters are passed a <emu-val>this</emu-val>
 value for the object on which property corresponding to the IDL attribute is
 accessed, they are able to expose instance-specific data.
 
-Note: Note that attempting to assign to a property corresponding to a
+Note: Attempting to assign to a property corresponding to a
```
Comment (tobie) [inline index.bs]: Can't/shouldn't we express that normatively?

File: index.bs [hunk 9]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
+                    |attribute|'s [=identifier=], and "getter".
+                1.  If |O| is not a [=platform object=] that implements the interface |target|,
+                    then:
+                    1.  If |attribute| was specified with the [{{LenientThis}}]
+                        [=extended attribute=], then return <emu-val>undefined</emu-val>.
+                    1. Otherwise, [=ECMAScript/throw=] a <emu-val>TypeError</emu-val>.
+            1.  Let |R| be the result of performing the actions listed in the
+                description of |attribute| that occur on getting (or those listed in the description
+                of the inherited attribute, if this attribute is declared to
+                [=inherit its getter=]), on |O| if |O| is not <emu-val>null</emu-val>.
```
Comment (tobie) [inline index.bs]: I find it hard to parse the meaning of the hanging conditional. I assume it means you carry out the getting in both case, but once it's carried with the |O| as `this` and the other one with `undefined` as `this`?

File: index.bs [hunk 10]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
+    The <dfn id="dfn-attribute-setter" export>attribute setter</dfn> is created as follows, given an
+    [=attribute=] |attribute|, an [=interface=] |target|, and a [=Realm=] |realm|:
+
+    1.  If |attribute| is [=read only=] and does not have a
+        [{{LenientThis}}], [{{PutForwards}}] or [{{Replaceable}}] [=extended attribute=], return
+        <emu-val>undefined</emu-val>; there is no [=attribute setter=] function.
+    1.  Assert: |attribute|'s type is not a [=promise type=].
+    1.  Let |steps| be the following series of steps:
+        1.  If no arguments were passed, then
+        1.  Let |V| be the value of the first argument passed.
```
Comment (tobie) [inline index.bs]: There's something unclear in the above two steps (either there's no argument or there is one).

File: index.bs [hunk 11]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
+        1.  Let |id| be |attribute|'s [=identifier=].
+        1.  Let |O| be <emu-val>null</emu-val>.
+        1.  If |attribute| is not a [=static attribute=]:
```
Comment (tobie) [inline index.bs]: `If |attribute| is a [=regular attribute=]:`

File: index.bs [hunk 12]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
+            1.  If the <emu-val>this</emu-val> value is <emu-val>null</emu-val> or
```
Comment (tobie) [inline index.bs]: If find all of the substeps of this step 5 super difficult to parse. I don't doubt that the algorithm is correct, but I really struggle to understand its intent. I can't imagine modifying something in here without accidentally breaking stuff.

Here's a proposed alternative (which is slightly less DRY, but hopefully a bit clearer). I added an optional note (prefixed with ?) to clarify the exit point of the substep. Maybe that helps (or is over the top—I'm not sure).

```
1.  If |attribute| is a [=regular attribute=]:
    1.  If the <emu-val>this</emu-val> value is <emu-val>null</emu-val> or
        <emu-val>undefined</emu-val>, set |O| to |realm|'s [=Realm/global object=].
        (This will subsequently cause a <emu-val>TypeError</emu-val> in a few steps, if
        the global object does not implement |target| and [{{LenientThis}}] is not
        specified.)
        <!-- https://www.w3.org/Bugs/Public/show_bug.cgi?id=18547#c9 -->
    1.  Otherwise, set |O| to the <emu-val>this</emu-val> value.
    1.  If |O| is a [=platform object=], then [=perform a security check=], passing |O|,
        |id|, and "setter".
     1.  Let |validThis| be `true` if |O| is a [=platform object=] that
         implements the interface |target|, or `false` otherwise.
     1.  If |validThis| is `true`, then:
         1.  If |attribute| is declared with a [{{LenientSetter}}] extended attribute, then:
             1.  Return <emu-val>undefined</emu-val>.
         1.  If |attribute| is declared with a [{{PutForwards}}] extended attribute, then:
             1.  Let |Q| be [=?=] [=Get=](|O|, |id|).
             1.  If [=Type=](|Q|) is not Object, then [=ECMAScript/throw=] a
                 <emu-val>TypeError</emu-val>.
             1.  Let |forwardId| be the identifier argument of the [{{PutForwards}}] extended
                 attribute.
             1.  Perform [=?=] [=Set=](|Q|, |forwardId|, |V|).
             1.  Return <emu-val>undefined</emu-val>.
         1.  If |attribute| is declared with the [{{Replaceable}}] extended attribute, then:
             1.  Perform [=?=] [=CreateDataProperty=](|O|, |id|, |V|).
             1.  Return <emu-val>undefined</emu-val>.
        
        ? Note: setters who are not special-cased here will
        ? continue being defined in the parent steps
        ? along with static attributes.
     1.  Otherwise if |attribute| was specified with the [{{LenientThis}}] [=extended attribute=],
         1.  If |attribute| is declared with the [{{Replaceable}}] extended attribute,
             then perform [=?=] [=CreateDataProperty=](|O|, |id|, |V|).
         1. Return <emu-val>undefined</emu-val>.
     1.  Otherwise, [=ECMAScript/throw=] a <emu-val>TypeError</emu-val>.
```

File: index.bs [hunk 13]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
+                <emu-val>undefined</emu-val>, set |O| to |realm|'s [=Realm/global object=].
+                (This will subsequently cause a <emu-val>TypeError</emu-val> in a few steps, if
+                the global object does not implement |target| and [{{LenientThis}}] is not
+                specified.)
+                <!-- https://www.w3.org/Bugs/Public/show_bug.cgi?id=18547#c9 -->
+            1.  Otherwise, set |O| to the <emu-val>this</emu-val> value.
+            1.  If |O| is a [=platform object=], then [=perform a security check=], passing |O|,
+                |id|, and "setter".
+            1.  Let |validThis| be <emu-val>true</emu-val> if |O| is a [=platform object=] that
```
Comment (tobie) [inline index.bs]: Shouldn't the booleans here really be just \`true\` ans \`false\`? These are not JS values, they're infra booleans, no?

File: index.bs [hunk 14]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
-            with |O| as the object and |idlValue| as the value.
-        1.  Otherwise, the attribute is a [=static attribute=].
-            Perform the actions listed in the description of the attribute that occur when setting with |idlValue| as the value.
-        1.  Return <emu-val>undefined</emu-val>.
-    </div>
+        1.  Let |idlValue| be determined as follows:
 
-    The value of the <emu-val>Function</emu-val> object’s “length”
-    property is the <emu-val>Number</emu-val> value <emu-val>1</emu-val>.
+            <dl class="switch">
```
Comment (tobie) [inline index.bs]: Why are we using a switch here rather than a regular if..otherwise clause?

File: index.bs [hunk 15]
```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
...
+                    |idlValue| is the result of [=converted to an IDL value|converting=] |V| to an
+                    IDL value of |attribute|'s type.
+                </dd>
+            </dl>
 
-    The value of the <emu-val>Function</emu-val> object’s “name”
-    property is a <emu-val>String</emu-val> whose value is the
-    concatenation of “set ” and the identifier of the attribute.
+        1.  Perform the actions listed in the description of |attribute| that occur on setting, on
+            |O| if |O| is not <emu-val>null</emu-val>.
```
Comment (tobie) [inline index.bs]: Same comments as for the hanging conditional in the getter, here.

**Comment by tobie** [comment]
A few nits and a few questions. LGTM otherwise.

File: index.bs (hunk 7 above)
Comment (domenic) [inline index.bs]: I'll go with not changing it for now as this area is related to the mixin mess and will get straightened out when that is.

File: index.bs (hunk 9 above)
Comment (domenic) [inline index.bs]: Well, there's no `this` formally (that's https://www.w3.org/Bugs/Public/show_bug.cgi?id=27301). The idea is not to provide anything at all; providing `undefined` isn't really correct.

File: index.bs (hunk 12 above)
Comment (domenic) [inline index.bs]: Hmm. I don't really find that clearer; I like how you just check the conditions one after another in the existing version. Let's leave it for a follow-up, as we'd also want to change the analogous steps for operations.

File: index.bs (hunk 13 above)
Comment (domenic) [inline index.bs]: That's probably more correct I guess. I was thinking this is more like an ES spec algorithm where booleans are emu-val'ed, but this seems like a better place to lean toward web spec conventions than ES.

File: index.bs (hunk 14 above)
Comment (domenic) [inline index.bs]: Not sure. We do it a lot, it seems. I guess because it's determining a value?

File: index.bs (hunk 15 above)
Comment (domenic) [inline index.bs]: I still think this is pretty clear as-is, but we could do a different version perhaps as a follow-up if you have a proposal.

File: index.bs (hunk 8 above)
Comment (domenic) [inline index.bs]: I'm pretty sure it is. This was just a drive-by fix to not repeat "Note" twice.

Comment (tobie) [inline index.bs]: For my own education, where is that behavior defined normatively?

**Comment by bzbarsky** [comment]
I'm really sorry for the horrible lag here.  I finally had a chance to read through these changes, and they look great.  There is one preexisting issue that I filed as https://github.com/heycam/webidl/issues/271

**Comment by bzbarsky** [comment]
Oh, and filed https://bugs.chromium.org/p/chromium/issues/detail?id=683310 for Blink not doing PutForwards per spec.

**Comment by bzbarsky** [comment]
And https://bugzilla.mozilla.org/show_bug.cgi?id=1332713 on updating Gecko to these changes.

**Comment by bzbarsky** [comment]
Also filed spec bugs on existing [SameObject] promises:

* https://github.com/w3c/presentation-api/issues/407
* https://github.com/w3c/ServiceWorker/issues/1060

**Comment by bzbarsky** [comment]
And https://github.com/w3c/testharness.js/pull/232 on idlharness no longer matching the spec now.  And filed https://github.com/w3c/testharness.js/issues/233 on what's left to do there.

**Comment by annevk** [comment]
It seems https://fetch.spec.whatwg.org/ was missed. It uses [SameObject] for trailers, a feature that isn't implemented yet.

**Comment by annevk** [comment]
Filed https://github.com/whatwg/fetch/issues/473.

**Comment by bzbarsky** [comment]
Right, I only filed bugs on the things I found by disallowing [SameObject] in Gecko's IDL and seeing what failed...
